libcxx/src/filesystem/operations.cpp

```diff
@@ -194,6 +271,46 @@ bool copy_file_impl(FileDescriptor& read_fd, FileDescriptor& write_fd, error_cod
 
   return true;
 }
+#endif
+
+#if defined(_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE) || defined(_LIBCPP_FILESYSTEM_USE_SENDFILE)
+bool copy_file_impl(FileDescriptor& read_fd, FileDescriptor& write_fd, error_code& ec) {
+#  if defined(_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE)
+  if (copy_file_impl_copy_file_range(read_fd, write_fd, ec)) {
+    return true;
+  }
+  // EINVAL: src and dst are the same file (this is not cheaply
+  // detectable from userspace)
+  // EINVAL: copy_file_range is unsupported for this file type by the
+  // underlying filesystem
+  // ENOTSUP: undocumented, can arise with old kernels and NFS
+  // EOPNOTSUPP: filesystem does not implement copy_file_range
+  // ETXTBSY: src or dst is an active swapfile (nonsensical, but allowed
+  // with normal copying)
+  // EXDEV: src and dst are on different filesystems that do not support
+  // cross-fs copy_file_range
+  // ENOENT: undocumented, can arise with CIFS
+  // ENOSYS: unsupported by kernel or blocked by seccomp
+  if (ec.value() != EINVAL && ec.value() != ENOTSUP && ec.value() != EOPNOTSUPP && ec.value() != ETXTBSY &&
+      ec.value() != EXDEV && ec.value() != ENOENT && ec.value() != ENOSYS) {
+    return false;
+  }
+  ec.clear();
+#  endif
+
+#  if defined(_LIBCPP_FILESYSTEM_USE_SENDFILE)
+  if (copy_file_impl_sendfile(read_fd, write_fd, ec)) {
+    return true;
+  }
+  // EINVAL: unsupported file type
+  if (ec.value() != EINVAL) {
+    return false;
+  }
+  ec.clear();
+#  endif
+
+  return copy_file_impl_fstream(read_fd, write_fd, ec);
+}
```

Comment:
I would rewrite this as follows, I think that makes it easier to understand and we duplicate very little code:

```c++
#if defined(_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE)

bool copy_file_impl(FileDescriptor& read_fd, FileDescriptor& write_fd, error_code& ec) {
  if (copy_file_impl_copy_file_range(read_fd, write_fd, ec)) {
    return true;
  }
  // EINVAL: src and dst are the same file (this is not cheaply
  // detectable from userspace)
  // EINVAL: copy_file_range is unsupported for this file type by the
  // underlying filesystem
  // ENOTSUP: undocumented, can arise with old kernels and NFS
  // EOPNOTSUPP: filesystem does not implement copy_file_range
  // ETXTBSY: src or dst is an active swapfile (nonsensical, but allowed
  // with normal copying)
  // EXDEV: src and dst are on different filesystems that do not support
  // cross-fs copy_file_range
  // ENOENT: undocumented, can arise with CIFS
  // ENOSYS: unsupported by kernel or blocked by seccomp
  if (ec.value() != EINVAL && ec.value() != ENOTSUP && ec.value() != EOPNOTSUPP && ec.value() != ETXTBSY &&
      ec.value() != EXDEV && ec.value() != ENOENT && ec.value() != ENOSYS) {
    return false;
  }
  ec.clear();

  return copy_file_impl_fstream(read_fd, write_fd, ec);
}

#elif defined(_LIBCPP_FILESYSTEM_USE_SENDFILE)

bool copy_file_impl(FileDescriptor& read_fd, FileDescriptor& write_fd, error_code& ec) {
  if (copy_file_impl_sendfile(read_fd, write_fd, ec)) {
    return true;
  }
  // EINVAL: unsupported file type
  if (ec.value() != EINVAL) {
    return false;
  }
  ec.clear();

  return copy_file_impl_fstream(read_fd, write_fd, ec);
}
 
#elif defined(_LIBCPP_FILESYSTEM_USE_COPYFILE)

...
```

And once you rewrite like that, you could also consider inlining `copy_file_impl_copy_file_range` and `copy_file_impl_sendfile` into their respective `copy_file_impl`. IDK if that would simplify the code but it might.

Reply:
That doesn't work, `copy_file_range` will (usually) fail on cross-fs copy, and we'd want to fall back to `sendfile` in that case, and not abandon ship and do a userspace copy.

Since only Linux and FreeBSD support `copy_file_range`, and only linux supports `sendfile` on files, perhaps something like:

```
#ifdef __linux__ || __FreeBSD__
copy_file_impl {
    #ifdef HAS_COPY_FILE_RANGE
    if(copy_file_range) { return; }
    #endif
    #ifdef HAS_SENDFILE
    if(sendfile) { return; }
    #endif
    fstream
}
#endif
```

Of course, this will get procedually more messy when another Unix platform implements any combination of `copy_file_range` and `sendfile`.

Reply:
There is something I don't understand. `_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE`, `_LIBCPP_FILESYSTEM_USE_SENDFILE` and `_LIBCPP_FILESYSTEM_USE_COPYFILE` are mutually exclusive (right?). My rewriting didn't change any logic at all, it only reorganized the code a bit to define the whole function inside each `#if` branch. Unless I missed something and my "refactoring" actually changed the logic?

Reply:
No, `_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE` and `_LIBCPP_FILESYSTEM_USE_SENDFILE` are not mutually exclusive. The intent is to attempt `copy_file_range`, and fall back to `sendfile` otherwise.

Reply:
Ah, I see what I missed now.

I would suggest something like this (which is basically what you wrote above IIUC):

```c++
#if HAS_SENDFILE
copy_file_impl_sendfile() { ... }
#endif

#if HAS_COPY_FILE_RANGE
copy_file_impl_copy_file_range() { ... }
#endif

#if HAS_COPY_FILE
copy_file_impl_copyfile() { ... }
#endif

copy_file_impl_fstream() { ... }


copy_file_impl() {
#if something
  full implementation
#elif something
  full implementation
#else
  full implementation
#endif
}
```

Does that make sense? Decouple the implementation of the different functions and the place where you compose them together, so that we can see at a glance what `copy_file_impl` does when looking at it.

Reply:
if you want the full implementation inlined into `copy_file_impl`, then what's the point of defining e.g. `copy_file_impl_sendfile` above?

Reply:
By "full implementation", I simply meant that you wouldn't interleave code within the `#ifdef` and code outside the `#ifdef`s, I didn't mean that you would inline everything recursively into `copy_file_impl`.

Reply:
is this to your liking now? The `error_code` handling still resides in `copy_file_impl`, as otherwise the various impls would have to return some kind of tribool to disambiguate soft failure due to preconditions and hard failure due to another error